mlflow/pyfunc/__init__.py

```diff
@@ -201,34 +201,70 @@ def spark_udf(spark, path, run_id=None, result_type="double"):
     Parameters passed to the UDF are forwarded to the model as a DataFrame where the names are
     ordinals (0, 1, ...).
 
+    If the model returns predictions as DataFrame, different result may be returned based on the
+    requested result_type. If the requested result type is numeric, the data frame is filtered to
+    contain only numeric columns first. If the requested return type is not an ArrayType, only the
+    first column will be returned.
+
     >>> predict = mlflow.pyfunc.spark_udf(spark, "/my/local/model")
     >>> df.withColumn("prediction", predict("name", "age")).show()
 
     :param spark: A SparkSession object.
     :param path: A path containing a :py:mod:`mlflow.pyfunc` model.
     :param run_id: ID of the run that produced this model. If provided, ``run_id`` is used to
                    retrieve the model logged with MLflow.
+    :param result_type: the return type of the user-defined function. The value can be either a
+                        :class:`pyspark.sql.types.DataType` object or a DDL-formatted type string.
     :return: Spark UDF type returned by the model's prediction method. Default double.
     """
 
     # Scope Spark import to this method so users don't need pyspark to use non-Spark-related
     # functionality.
     from mlflow.pyfunc.spark_model_cache import SparkModelCache
     from pyspark.sql.functions import pandas_udf
+    from pyspark.sql.types import _parse_datatype_string
+    from pyspark.sql.types import ArrayType, NumericType, StringType, DataType
+
+    if not isinstance(result_type, DataType):
+        result_type = _parse_datatype_string(result_type)
+
+    elem_type = result_type
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+    if not isinstance(elem_type, NumericType) and not isinstance(elem_type, StringType):
+        raise Exception("Invalid requested type ''. pyfunc.spark_udf currently only supports " +
```

Comment:
We should raise an `MlflowException` here with an appropriate error code from the `mlflow.protos.databricks_pb2` module.